snuba/query/parser/__init__.py

```diff
@@ -373,6 +374,38 @@ def _expand_aliases(query: Query) -> None:
 ARRAYJOIN_FUNCTION_MATCH = FunctionCallMatch(StringMatch("arrayJoin"), None)
 
 
+def _mangle_aliases(query: Query) -> None:
+    alias_prefix = "_snuba_"
+
+    def transform_alias(alias: Optional[str]) -> Optional[str]:
+        if alias is not None:
+            return f"{alias_prefix}{alias}"
+
+        return alias
+
+    query.transform_expressions(
+        lambda expr: replace(expr, alias=transform_alias(expr.alias))
+    )
+
+    for expression in query.get_all_expressions():
+        if expression.alias:
+            expression.transform(
+                lambda expr: replace(expr, alias=f"{alias_prefix}{expr.alias}")
+            )
+
+    # HACK: This reverts the mangle of arrayjoin since we cannot reference aliases
+    # there
```

Comment:
Not following this. What do you mean by "we cannot reference aliases there"

Reply:
I mean that ClickHouse will not run `SELECT exception_stacks.type AS alias FROM sentry_local ARRAY JOIN alias`, so I've rewritten into `SELECT exception_stacks.type AS alias FROM sentry_local ARRAY JOIN exception_stacks.type`

Reply:
This works though.

```
SELECT 
    event_id, 
    aaa
FROM sentry_local
ARRAY JOIN tags.key AS aaa

┌─event_id─────────────────────────┬─aaa────┐
│ 855f59b850844461b48349c769470b42 │ color  │
│ 855f59b850844461b48349c769470b42 │ level  │
│ 855f59b850844461b48349c769470b42 │ paying │
└──────────────────────────────────┴────────┘
┌─event_id─────────────────────────┬─aaa────┐
│ 17315b86f5af4859b30e46c2262b993b │ color  │
│ 17315b86f5af4859b30e46c2262b993b │ level  │
│ 17315b86f5af4859b30e46c2262b993b │ paying │
└──────────────────────────────────┴────────┘

6 rows in set. Elapsed: 0.006 sec. 

be29e23e76c8 :) 
```

I think we have three issues here:
1) We could be transforming the column we are using for the `ARRAY JOIN` in several places in the query, but the formatter (which actually assigns aliases) does not care of formatting the ARRAY JOIN clause before everything else so it won't depend on an alias. This is independent from your PR. Could you track this in the backlog ?
2) We are allowing any expression in the AST to be used in the ARRAY JOIN clause. Not a big deal since the language only accepts columns there.
3) In the previous system we were applying aliases to all columns (the alias was the same string as the column name, which were removed when possible during formatting). This is not needed anymore if we do not need to preserve aliases that the client did not provide and since we expand all of them. I think you should not be mangle column aliases when they are the same as the name. You should be able to just remove them. Which basically means undoing what `_apply_column_aliases` does, I don't think you can just skip that step since alias expanding depends on it, though you should be able to remove those that are not needed. This would generate less verbose queries (with fewer chances to fail for their size) and would keep ARRAY JOIN consistent with rest of the query.